odoo/orm/fields.py

```diff
@@ -1107,6 +1107,207 @@ def update_db_related(self, model):
+            sql_expr = self._condition_to_sql_company(sql_expr, model, alias, field_expr, operator, value, query)
+        return sql_expr
+
+    def _condition_to_sql_company(self, sql_expr: SQL, model: BaseModel, alias: str, field_expr: str, operator: str, value, query: Query) -> SQL:
+        """ Add a not null condition on the field for company-dependent fields to use an existing index for better performance."""
+        if (
+            self.company_dependent
+            and self.index == 'btree_not_null'
+            and not (self.type in ('datetime', 'date') and field_expr != self.name)  # READ_GROUP_NUMBER_GRANULARITY is not supported
+            and model.env['ir.default']._evaluate_condition_with_fallback(model._name, field_expr, operator, value) is False
+        ):
+            return SQL('(%s IS NOT NULL AND %s)', SQL.identifier(alias, self.name), sql_expr)
+        return sql_expr
+
+    @classmethod
+    def _convert_sql_to_sql_text(cls, sql_field: SQL) -> SQL:
+        """ Cast the field to char."""
+        return SQL("%s::text", sql_field)
+
+    def __condition_to_sql(self, model: BaseModel, alias: str, field_expr: str, operator: str, value, query: Query) -> SQL:
+        sql_field = model._field_to_sql(alias, field_expr, query)
+
+        def _value_to_column(v):
+            return self.convert_to_column(v, model, validate=False)
+
+        # -------------------------------------------------
+        # support for SQL value
+        # TODO deprecate this usage
+        if operator in SQL_OPERATORS and isinstance(value, SQL):
+            return SQL("%s%s%s", sql_field, SQL_OPERATORS[operator], value)
+
+        # -------------------------------------------------
+        # operator: in (equality)
+        equal_operator = None
+        if operator in ('=', '!='):
+            equal_operator = operator
+            operator = 'in' if operator == '=' else 'not in'
+            value = [value]
+        if operator in ('in', 'not in'):
+            assert isinstance(value, COLLECTION_TYPES), \
+                f"condition_to_sql() 'in' operator expects a collection, not a {value!r}"
+            if not value:
+                return SQL("FALSE") if operator == 'in' else SQL("TRUE")
+            if self.name == 'id':
+                params = tuple(v for v in value if v)
+            elif self.relational:
+                params = tuple(_value_to_column(v) for v in value if v)
+            else:
+                params = tuple(_value_to_column(v) for v in value if v is not False and v is not None)
+            null_in_condition = len(params) < len(value)
+            # if we have a value treated as null
+            if (null_value := self.falsy_value) is not None:
+                null_value = _value_to_column(null_value)
+                if null_value in params:
+                    null_in_condition = True
+                elif null_in_condition:
+                    params = (*params, null_value)
```

Comment:
This whole mess could be part of my proposed method `_prepare_in_values` above, and only for the fields that need it.

Reply:
If it would be defined in an overridden function, the logic would just be duplicated for each field having falsy values.